File: Src/Engine/Editor/GUI/Input/SearchBox.cpp

#include "SearchBox.h"

#include "Editor/EditorIcons.h"
#include "Runtime/UI/GUI/Style.h"
#include "Runtime/UI/GUI/Brushes/SpriteBrush.h"
#include "Runtime/UI/GUI/Common/Button.h"

namespace SE::Editor
{
	SearchBox::SearchBox() : TextBox(false, 0, 0), ClearSearchButton(nullptr)
	{
	}

	SearchBox::SearchBox(bool isMultiline, float x, float y, float width): TextBox(isMultiline, x, y, width)
	{
		WatermarkText = SE_TEXT("Search...");

		ClearSearchButton = New<Button>();
		ClearSearchButton->Parent = this;
		ClearSearchButton->Width = 14.0f;
		ClearSearchButton->Height = 14.0f;
		ClearSearchButton->AnchorPreset = AnchorPresets::TopRight;
		ClearSearchButton->Text = SE_TEXT("");
		ClearSearchButton->TooltipText = SE_TEXT("Cancel Search.");
		ClearSearchButton->BackgroundColor = TextColor;
		ClearSearchButton->BorderColor = Colors::Transparent;
		ClearSearchButton->BackgroundColorHighlighted = Style::Current->ForegroundGrey;
		ClearSearchButton->BorderColorHighlighted = Colors::Transparent;
		ClearSearchButton->BackgroundColorSelected = Style::Current->ForegroundGrey;
		ClearSearchButton->BorderColorSelected = Colors::Transparent;
		ClearSearchButton->BackgroundBrush = New<SpriteBrush>(EditorIcons::Cross16);
		ClearSearchButton->Visible = false,

		ClearSearchButton->LocalY = ClearSearchButton->LocalY + 2;
		ClearSearchButton->LocalX = ClearSearchButton->LocalX - 2;
		ClearSearchButton->Clicked.Bind([this]
		{
			Clear();
		});
		ClearSearchButton->HoverBegin.Bind([this]
		{
			_changeCursor = false;
			Cursor = CursorType::Default;
		});
		ClearSearchButton->HoverEnd.Bind([this]
		{
			_changeCursor = true;
		});

		TextChanged.Bind([this]
		{
			ClearSearchButton->Visible = !Text.operator->().IsEmpty();
		});
	}
} // SE

File: Src/Engine/Editor/GUI/Input/SearchBox.cs
using System;
using SE.GUI;

namespace SE.Editor.GUI
{
    public sealed class SearchBox : TextBox
    {
        public SearchBox()
            : this(false, 0.0f, 0.0f)
        {
        }

        public SearchBox(bool isMultiline, float x, float y, float width = 120.0f)
            : this(x, y, width, isMultiline)
        {
        }

        public SearchBox(float x = 0, float y = 0, float width = 120, bool isMultiline = false)
            : base(new Rectangle(x, y, width, 22), isMultiline)
        {
            PlaceholderText = "Search...";
            ClearSearchButton = new Button(new Rectangle(width - 18, 4, 14, 14), "x");
            ClearSearchButton.Visible = false;
            ClearSearchButton.TooltipText = "Cancel Search.";
            ClearSearchButton.Clicked += _ => Clear();

            AddChild(ClearSearchButton);
        }

        public event Action<SearchBox>? TextChanged;

        public Button ClearSearchButton { get; }
        public string WatermarkText
        {
            get => PlaceholderText;
            set => PlaceholderText = value;
        }

        public void SetText(string text)
        {
            Text = text;
        }

        public void Clear()
        {
            SetText(string.Empty);
        }

        protected override void OnTextChanged()
        {
            base.OnTextChanged();
            ClearSearchButton.Visible = !string.IsNullOrEmpty(Text);
            ClearSearchButton.SetBounds(Width - 18, 4, 14, 14);
            TextChanged?.Invoke(this);
        }
    }
}
